staging/src/k8s.io/apiserver/pkg/server/genericapiserver_graceful_termination_test.go

```diff
@@ -429,58 +535,252 @@ func TestMuxAndDiscoveryComplete(t *testing.T) {
+		// wait until the test commands this pre shutdown
+		// hook to invoke an API call.
+		<-hook.blockedCh
+
+		resultGot := doer.Do(client, func(httptrace.GotConnInfo) {}, "/echo?message=request-from-pre-shutdown-hook-should-succeed", time.Second)
+		hook.resultCh <- resultGot
+		return nil
+	}); err != nil {
+		t.Fatalf("Failed to register preshutdown hook - %v", err)
+	}
+
+	return hook
+}
+
+type fakeAudit struct {
+	shutdownCh chan struct{}
+	lock       sync.Mutex
+	audits     map[string]struct{}
+}
+
+func (a *fakeAudit) Run(stopCh <-chan struct{}) error {
+	a.shutdownCh = make(chan struct{})
 	go func() {
-		defer func() {
-			// this test has an inherent race condition when we wait for two go routines
-			// to finish - the Run method and the pre-shutdown hook, each running in
-			// its own goroutine, give it a second before unblocking the test assert
-			<-time.After(time.Second)
-			close(runCompletedCh)
-		}()
-		s.PrepareRun().Run(stopCh)
+		defer close(a.shutdownCh)
+		<-stopCh
 	}()
-	waitForAPIServerStarted(t, doer)
+	return nil
+}
 
-	close(stopCh)
+func (a *fakeAudit) Shutdown() {
+	<-a.shutdownCh
+}
+
+func (a *fakeAudit) String() string {
+	return "fake-audit"
+}
 
+func (a *fakeAudit) ProcessEvents(events ...*auditinternal.Event) bool {
+	a.lock.Lock()
+	defer a.lock.Unlock()
+	if len(a.audits) == 0 {
+		a.audits = map[string]struct{}{}
+	}
+	for _, event := range events {
+		a.audits[string(event.AuditID)] = struct{}{}
+	}
+
+	return true
+}
+
+func (a *fakeAudit) requestAudited(auditID string) bool {
+	a.lock.Lock()
+	defer a.lock.Unlock()
+	_, exists := a.audits[auditID]
+	return exists
+}
+
+func (a *fakeAudit) EvaluatePolicyRule(attrs authorizer.Attributes) audit.RequestAuditConfigWithLevel {
+	return audit.RequestAuditConfigWithLevel{
+		Level:              auditinternal.LevelMetadata,
+		RequestAuditConfig: audit.RequestAuditConfig{},
+	}
+}
+
+func assertRequestAudited(resultGot result, backend *fakeAudit) error {
+	resp := resultGot.response
+	if resp == nil {
+		return fmt.Errorf("Expected a response, but got nil")
+	}
+	auditIDGot := resp.Header.Get(auditinternal.HeaderAuditID)
+	if len(auditIDGot) == 0 {
+		return fmt.Errorf("Expected non-empty %q response header, but got: %v", auditinternal.HeaderAuditID, resp)
+	}
+	if !backend.requestAudited(auditIDGot) {
+		return fmt.Errorf("Expected the request to be audited: %q", auditIDGot)
+	}
+	return nil
+}
+
+func waitForeverUntil(t *testing.T, ch <-chan struct{}, msg string) {
 	select {
-	case err := <-preShutdownHookErrCh:
-		if err != nil {
-			t.Fatalf("PreSHutdown hook can not access the API server - %v", err)
-		}
-	case <-runCompletedCh:
-		t.Fatalf("API Server exited without running the PreShutdown hooks")
-	case <-time.After(15 * time.Second):
-		t.Fatalf("test timed out after 15 seconds")
+	case <-ch:
+	case <-time.After(wait.ForeverTestTimeout):
+		t.Fatalf("%s", msg)
+	}
+}
+
+func waitForeverUntilSignaled(t *testing.T, s lifecycleSignal) {
```

Comment:
Do we need it? Can't we just simply call `waitForeverUntil(t, ch.Signaled()` where needed?

Reply:
it's a passthrough to `waitForeverUntil`, just want to avoid saying `waitForeverUntil(t, s.Signaled(), fmt.Sprintf("Expected the server to signal %s event", s.Name()))` over and over